### ISSUE TITLE
[Bug] Injecting service not working with async variable (pool asyncpg)

### ISSUE THREAD

**Issue by BenJeau** [sanic-org/sanic-ext#112]
**Describe the bug**
Unable to extract `asyncpg` Pool (https://magicstack.github.io/asyncpg/current/api/index.html#connection-pools) when added as a dependency, `None` is returned even though it is defined in the `before_server_start` listener (this is shown in the logs below).

Is there something special that needs to be done for async related variables? Aside from that, really love the injections/dependencies (useful for extracting user from bearer token)!

**Screenshots**
```sh
[2022-08-06 11:14:42 -0400] [60755] [INFO] 
  ┌───────────────────────────────────────────────────────────────────────────────┐
  │                                 Sanic v22.3.2                                 │
  │                        Goin' Fast @ http://0.0.0.0:8888                       │
  ├───────────────────────┬───────────────────────────────────────────────────────┤
  │                       │        mode: debug, single worker                     │
  │     ▄███ █████ ██     │      server: sanic                                    │
  │    ██                 │      python: 3.10.4                                   │
  │     ▀███████ ███▄     │    platform: macOS-12.5-arm64-arm-64bit               │
  │                 ██    │ auto-reload: enabled                                  │
  │    ████ ████████▀     │    packages: sanic-routing==22.3.0, sanic-ext==22.3.2 │
  │                       │                                                       │
  │ Build Fast. Run Fast. │                                                       │
  └───────────────────────┴───────────────────────────────────────────────────────┘

[2022-08-06 11:14:42 -0400] [60755] [INFO] Sanic Extensions:
[2022-08-06 11:14:42 -0400] [60755] [INFO]   > http 
[2022-08-06 11:14:42 -0400] [60755] [INFO]   > openapi [http://0.0.0.0:8888/docs]
[2022-08-06 11:14:42 -0400] [60755] [INFO]   > injection [0 added]
[2022-08-06 11:14:42 -0400] [60755] [DEBUG] Dispatching signal: server.init.before
pool before server start <asyncpg.pool.Pool object at 0x10230d630>
test_class before server start <__main__.TestClass object at 0x1024b6b60>
[2022-08-06 11:14:42 -0400] [60755] [DEBUG] Dispatching signal: server.init.after
[2022-08-06 11:14:42 -0400] [60755] [INFO] Starting worker [60755]
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.begin
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.read_head
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.request
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.handle
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.routing.before
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.routing.after
pool in handler None
test_class in handler <__main__.TestClass object at 0x1024b6b60>
request.app.ctx namespace(_dependencies=namespace(pool=<asyncpg.pool.Pool object at 0x10230d630>, test_class=<__main__.TestClass object at 0x1024b6b60>), cors=CORSSettings(allow_headers=frozenset({'*'}), allow_methods=frozenset(), allow_origins=(), always_send=True, automatic_options=True, expose_headers=frozenset(), max_age='5', send_wildcard=False, supports_credentials=False))
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.middleware.before
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.middleware.after
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.response
[2022-08-06 11:14:55 -0400] - (sanic.access)[INFO][127.0.0.1:59983]: GET http://0.0.0.0:8888/test  200 2
[2022-08-06 11:14:55 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.send
[2022-08-06 11:15:00 -0400] [60755] [DEBUG] KeepAlive Timeout. Closing connection.
[2022-08-06 11:15:00 -0400] [60755] [DEBUG] Dispatching signal: http.lifecycle.complete
```

**To Reproduce**
Here's a code sample, you'll probably need to connect to a postgresql database to fully test it (replacing that "psql database url" string).

```py
import asyncio
from asyncpg import Pool, create_pool
from sanic import Sanic, Request, text
from sanic_ext import Extend


app = Sanic("test")
Extend(app)


class TestClass:
    def execute(self, query: str) -> str:
        return query


@app.get("/test")
def handler_test(request: Request, pool: Pool, test_class: TestClass):
    print("pool in handler", pool)
    print("test_class in handler", test_class)
    print("request.app.ctx", request.app.ctx)

    return text("ok")


async def t():
    await asyncio.sleep(5)
    return TestClass()


@app.before_server_start
async def setup_db(app: Sanic, loop):
    pool = await create_pool("psql database url", loop=loop, max_size=100)
    print("pool before server start", pool)
    app.ext.dependency(pool)

    test_class = await t()
    print("test_class before server start", test_class)
    app.ext.dependency(test_class)


if __name__ == "__main__":
    app.run(host="0.0.0.0",
            port=8888,
            dev=True,
            fast=False)
```

**Expected behavior**
Be able to extract `pool` successfully just like the `test_class`

**Environment (please complete the following information):**
 - OS: macOS-12.5-arm64-arm-64bit
 - Browser: Postman + chrome
 - Version: 22.3.2

> __Edit__: changed the test with TestClass to be received from an async function, so maybe it's not related to that, I'm unsure

**Comment by BenJeau** [comment]
The interesting thing is that you see it in the `request.app.ctx` in `_dependencies`, so maybe I'm maybe missing something at the handler level or there's potentially a bug? Thanks in advance :)

**Comment by BenJeau** [comment]
It seems like the main difference is the `TestClass` is of type `__main__.TestClass` whereas the `Pool` is just `Pool`. Unsure if there's a way to get the `__main__` from an instance of a class (if that's what is required)

**Comment by BenJeau** [comment]
Looking at bit more at the code, it seems like it's related to the `do_cast` function in the `constructor.py` file of the `injection` folder - although need to debug it more to know how it works, it seems to be related to the way it's calling the constructor? If anyone has more info, please let me know here :)

**Comment by ahopkins** [comment]
Not at a computer to check, but I've definitely had this working with `asyncpg`.

My initial thought is maybe

```python
asyncpg.Pool is not async.pool.Pool
```

I can take a look later today though.

**Comment by BenJeau** [comment]
I did try that, as I thought the same - I'm also not close to a computer, but can double check. My conclusion yesterday seemed that I need to create a class that wrapped asyncpg, although it's a bit gross to do imo, but if you say you've got it working, I may be doing something wrong.

**Comment by ahopkins** [comment]
You absolutely should not need to wrap the class. Will try this evening.

**Comment by BenJeau** [comment]
Just tried it with `async.pool.Pool` and it doesn't work - and thanks for your help :)

**Comment by ahopkins** [comment]
Oh my.... I see the problem. :thinking: 

```python
>>> import asyncio
>>> from asyncpg import create_pool
>>> pool = await create_pool("<DSN>")
>>> from inspect import isawaitable
>>> isawaitable(pool)
True
```

The `Pool` object is awaitable...

![image](https://user-images.githubusercontent.com/166269/183310755-9ec63840-4cdd-4774-8daf-a7b4181c8ab5.png)

This should probably be `iscoroutine` instead.

**Comment by BenJeau** [comment]
Oh that's interesting, not an expert on Python's asyncs, but from your observation, I've changed the usage of `isawaitable` to `iscoroutine` here:

https://github.com/sanic-org/sanic-ext/blob/b65cd9f28c8d1e6ee0aad91685bfb84fc4925ac6/sanic_ext/extensions/injection/constructor.py#L47

and I'm now getting the pool in the handler. I can open a PR for that unless you are already in the process of doing so or have other changes in mind related to this.

**Comment by ahopkins** [comment]
Sweet, please do. If you can make a quick dummy object that causes the bahavior and thrown in a test that would be awesome also. I want to get this and some other bug fixes released this week.

**Comment by ahopkins** [comment]
Resolved by #121
